File: engine/src/main/java/franca/java/graphics/animations/LinearEase.java
package franca.java.graphics.animations;

import franca.java.graphics.device.Device;

public class LinearEase extends Ease {

  public LinearEase(int initialValue, int targetValue, long duration) {
    super(initialValue, targetValue, duration);
  }

  @Override
  public void destroy() {
    super.destroy();
  }

  @Override
  public boolean tick(Device device, long startedTime) {

    // returns true when needs repainting

    if (currentValue == targetValue) {
      return false;
    }

    long currentTime = device.getTime();

    if (currentTime >= startedTime + duration) {
      currentValue = targetValue;
      return true;
    }

    currentValue = (int) (initialValue + (targetValue - initialValue) * (currentTime - startedTime) / duration);
    return true;
  }
}

File: platforms/typescript/src/java-franca/animations/linear-ease.ts
import {Ease} from "./ease";
import {Router} from "../device/router";

export class LinearEase extends Ease {
  constructor(initialValue: number, targetValue: number, duration: number) {
    super(initialValue, targetValue, duration);
  }

  destroy(): void {
    super.destroy();
  }

  tick(router: Router, startedTime: number): boolean {
    // returns true when needs repainting

    if (this.currentValue === this.targetValue) {
      return false;
    }

    const currentTime = router.getDevice()!.getTime();

    if (currentTime >= startedTime + this.duration) {
      this.currentValue = this.targetValue;
      return true;
    }

    this.currentValue = Math.floor(
      this.initialValue +
      (this.targetValue - this.initialValue) *
      (currentTime - startedTime) /
      this.duration
    );
    return true;
  }
}
